File: api/getUserTweets.ts
import { get } from "../utils/request";

const url = 'https://x.com/i/api/graphql/il0axqlKn9gdWoOuhrfLbQ/UserTweets';

export async function getUserTweets(userId: string, count: number = 20) {
    const params = {
        variables: {
            "userId": userId,
            "count": count,
            "includePromotedContent":true,
            "withQuickPromoteEligibilityTweetFields":true,
            "withVoice":true,
            "withV2Timeline":true
        },
        features: {
            "profile_label_improvements_pcf_label_in_post_enabled":true,
            "rweb_tipjar_consumption_enabled":true,
            "responsive_web_graphql_exclude_directive_enabled":true,
            "verified_phone_label_enabled":false,
            "creator_subscriptions_tweet_preview_api_enabled":true,
            "responsive_web_graphql_timeline_navigation_enabled":true,
            "responsive_web_graphql_skip_user_profile_image_extensions_enabled":false,
            "premium_content_api_read_enabled":false,
            "communities_web_enable_tweet_community_results_fetch":true,
            "c9s_tweet_anatomy_moderator_badge_enabled":true,
            "responsive_web_grok_analyze_button_fetch_trends_enabled":false,
            "responsive_web_grok_analyze_post_followups_enabled":true,
            "responsive_web_jetfuel_frame":false,
            "responsive_web_grok_share_attachment_enabled":true,
            "articles_preview_enabled":true,
            "responsive_web_edit_tweet_api_enabled":true,
            "graphql_is_translatable_rweb_tweet_is_translatable_enabled":true,
            "view_counts_everywhere_api_enabled":true,
            "longform_notetweets_consumption_enabled":true,
            "responsive_web_twitter_article_tweet_consumption_enabled":true,
            "tweet_awards_web_tipping_enabled":false,
            "creator_subscriptions_quote_tweet_preview_enabled":false,
            "freedom_of_speech_not_reach_fetch_enabled":true,
            "standardized_nudges_misinfo":true,
            "tweet_with_visibility_results_prefer_gql_limited_actions_policy_enabled":true,
            "rweb_video_timestamps_enabled":true,
            "longform_notetweets_rich_text_read_enabled":true,
            "longform_notetweets_inline_media_enabled":true,
            "responsive_web_grok_image_annotation_enabled":true,
            "responsive_web_enhance_cards_enabled":false},
        fieldToggles: {
            "withArticlePlainText":false
        }
    };
    return await get(url, params);
};

File: api/get_user_tweets.py
import json
import requests
from util.request import request
from util.get_cursor_from_instructions import get_cursor_from_instructions
from typing import Tuple

url = 'https://x.com/i/api/graphql/il0axqlKn9gdWoOuhrfLbQ/UserTweets'

def get_user_tweets(session: requests.Session, user_id: str, count: int = 20, cursor: str = None) -> Tuple[dict, str]:
    """
    Get tweets of a user

    rate limit: 50/min
    """
    variables = {
        "userId": user_id,
        "count": count,
        "includePromotedContent": False,
        "withQuickPromoteEligibilityTweetFields": True,
        "withVoice": True,
        "withV2Timeline": True
    }
    if cursor:
        variables["cursor"] = cursor
    params = {
        "variables": json.dumps(variables),
        "features": json.dumps({
            "profile_label_improvements_pcf_label_in_post_enabled": True,
            "rweb_tipjar_consumption_enabled": True,
            "responsive_web_graphql_exclude_directive_enabled": True,
            "verified_phone_label_enabled": False,
            "creator_subscriptions_tweet_preview_api_enabled": True,
            "responsive_web_graphql_timeline_navigation_enabled": True,
            "responsive_web_graphql_skip_user_profile_image_extensions_enabled": False,
            "premium_content_api_read_enabled": False,
            "communities_web_enable_tweet_community_results_fetch": True,
            "c9s_tweet_anatomy_moderator_badge_enabled": True,
            "responsive_web_grok_analyze_button_fetch_trends_enabled": False,
            "responsive_web_grok_analyze_post_followups_enabled": True,
            "responsive_web_jetfuel_frame": False,
            "responsive_web_grok_share_attachment_enabled": True,
            "articles_preview_enabled": True,
            "responsive_web_edit_tweet_api_enabled": True,
            "graphql_is_translatable_rweb_tweet_is_translatable_enabled": True,
            "view_counts_everywhere_api_enabled": True,
            "longform_notetweets_consumption_enabled": True,
            "responsive_web_twitter_article_tweet_consumption_enabled": True,
            "tweet_awards_web_tipping_enabled": False,
            "creator_subscriptions_quote_tweet_preview_enabled": False,
            "freedom_of_speech_not_reach_fetch_enabled": True,
            "standardized_nudges_misinfo": True,
            "tweet_with_visibility_results_prefer_gql_limited_actions_policy_enabled": True,
            "rweb_video_timestamps_enabled": True,
            "longform_notetweets_rich_text_read_enabled": True,
            "longform_notetweets_inline_media_enabled": True,
            "responsive_web_grok_image_annotation_enabled": True,
            "responsive_web_enhance_cards_enabled": False
        }),
        "fieldToggles": json.dumps({
            "withArticlePlainText": False
        })
    }
    data = request(session, url, params=params)
    instructions = data["data"]["user"]["result"]["timeline_v2"]["timeline"]["instructions"]
    next_cursor = get_cursor_from_instructions(instructions)
        
    return data, next_cursor
